Maintenance and Operation full CRUD

sbriclot opened the pull request:
Maintenance#index includes the form to add/edit a maintenance.
Operation#index includes the form to add/edit an operation.
An operation can be link to a task from the maintenance plan or not.
When a MP is selected in the list, task and details of the operation are populated, and the default values can be updated if required.

Add a link from vehicle#show to maintenance#index.

sbriclot committed:
Maintenance and Operation full CRUD

JPS09 commented on app/controllers/maintenances_controller.rb:
```diff
@@ -1,2 +1,69 @@
+    @maintenance.vehicle_id = params[:vehicle_id]
+
+    if @maintenance.save
+      redirect_to vehicle_maintenance_operations_path(params[:vehicle_id], @maintenance)
```
I think that you could use the instance variable @vehicle instead of digging into the params since you used a method to get the desired vehicle.
```suggestion
    @maintenance.vehicle_id = @vehicle

    if @maintenance.save
      redirect_to vehicle_maintenance_operations_path(@vehicle, @maintenance)
```

JPS09 commented on app/controllers/maintenances_controller.rb:
```diff
@@ -1,2 +1,69 @@
+  def update
+    @maintenance.vehicle_id = params[:vehicle_id]
+    if @maintenance.update(maintenance_params)
+      redirect_to vehicle_maintenances_path(params[:vehicle_id])
```
Same thing here :) 
```suggestion
  def update
    @maintenance.vehicle_id = @vehicle
    if @maintenance.update(maintenance_params)
      redirect_to vehicle_maintenances_path(@vehicle)
```

JPS09 commented on app/controllers/operations_controller.rb:
```diff
@@ -1,2 +1,78 @@
+
+  def create
+    @operation = Operation.new(operation_params)
+    @operation.maintenance_id = params[:maintenance_id]
```
Same idea here
```suggestion
    @operation.maintenance_id = @maintenance
```

JPS09 commented on app/controllers/operations_controller.rb:
```diff
@@ -1,2 +1,78 @@
+  def update
+    @operation.maintenance_plan_id = extract_mp_id
+    if @operation.update(operation_params)
+      redirect_to vehicle_maintenance_operations_path(params[:vehicle_id], params[:maintenance_id])
```
Also here, is there a particular reason of using params for this CRUD?
```suggestion
      redirect_to vehicle_maintenance_operations_path(@vehicle, @maintenance)
```

JPS09 commented on app/controllers/operations_controller.rb:
```diff
@@ -1,2 +1,78 @@
+  def destroy
+    @operation.destroy
+
+    redirect_to vehicle_maintenance_operations_path(params[:vehicle_id], params[:maintenance_id])
```
```suggestion
    redirect_to vehicle_maintenance_operations_path(@vehicle, @maintenance)
```

JPS09 commented on app/views/maintenance_plans/index.html.erb:
```diff
@@ -35,3 +35,7 @@
 </div>
+
+<div>
+  <%= link_to "Retour", vehicle_path(params[:vehicle_id]) %>
```
I think that you have access to the the @vehicle variable here
```suggestion
  <%= link_to "Retour", vehicle_path(@vehicle) %>
```

JPS09 commented on app/views/maintenances/index.html.erb:
```diff
@@ -0,0 +1,26 @@
+          <%= maintenance.date.strftime("%m/%d/%Y") %>
+          <%= "#{maintenance.km} km" %>
+          <%= pluralize(maintenance.nb_ops, "opération") %>
+          <%= link_to "Details", vehicle_maintenance_operations_path(params[:vehicle_id], maintenance) %>
+          | 
+          <%= link_to "Edit", edit_vehicle_maintenance_path(params[:vehicle_id], maintenance) %>
+          | 
+          <%= link_to "Del.", vehicle_maintenance_path(params[:vehicle_id], maintenance), method: :delete, remote: true %>
```
```suggestion
          <%= link_to "Details", vehicle_maintenance_operations_path(@vehicle, maintenance) %>
          | 
          <%= link_to "Edit", edit_vehicle_maintenance_path(@vehicle, maintenance) %>
          | 
          <%= link_to "Del.", vehicle_maintenance_path(@vehicle, maintenance), method: :delete, remote: true %>
```

JPS09 replied:
I think you could replace these params as well

JPS09 commented on app/views/maintenances/index.html.erb:
```diff
@@ -0,0 +1,26 @@
+</div>
+
+<div>
+  <%= link_to "Retour", vehicle_path(params[:vehicle_id]) %>
```
```suggestion
  <%= link_to "Retour", vehicle_path(@vehicle) %>
```

JPS09 replied:
Same here

JPS09 commented on app/views/operations/index.html.erb:
```diff
@@ -0,0 +1,25 @@
+        <div class="list-el">
+          <div><%= operation.task %></div>
+          <div><%= operation.details %></div>
+          <%= link_to "Edit", edit_vehicle_maintenance_operation_path(params[:vehicle_id], operation.maintenance_id, operation) %>
+          | 
+          <div><%= link_to "Del.", vehicle_maintenance_operation_path(params[:vehicle_id], operation.maintenance_id, operation), method: :delete, remote: true %></div>
```
```suggestion
          <%= link_to "Edit", edit_vehicle_maintenance_operation_path(@vehicle, operation.maintenance_id, operation) %>
          | 
          <div><%= link_to "Del.", vehicle_maintenance_operation_path(@vehicle, operation.maintenance_id, operation), method: :delete, remote: true %></div>
```

JPS09 replied:
You have access to @vehicle here

JPS09 commented on app/views/operations/index.html.erb:
```diff
@@ -0,0 +1,25 @@
+</div>
+
+<div>
+  <%= link_to "Retour", vehicle_maintenances_path(params[:vehicle_id]) %>
```
```suggestion
  <%= link_to "Retour", vehicle_maintenances_path(@vehicle) %>
```

JPS09 replied:
Same here

JPS09 commented on app/controllers/operations_controller.rb:
```diff
@@ -1,2 +1,78 @@
+    @operation.maintenance_plan_id = extract_mp_id
+
+    if @operation.save
+      redirect_to vehicle_maintenance_operations_path(params[:vehicle_id], params[:maintenance_id])
```
```suggestion
      redirect_to vehicle_maintenance_operations_path(@vehicle, @maintenance)
```

JPS09 commented on app/controllers/maintenances_controller.rb:
```diff
@@ -1,2 +1,69 @@
 class MaintenancesController < ApplicationController
+  before_action :set_maintenance, only: %i[show edit update destroy]
+  before_action :set_vehicle, only: %i[index create edit update]
```
Allows to use @vehicle in destroy for redirection
```suggestion
  before_action :set_vehicle, only: %i[index create edit update destroy]
```

JPS09 commented on app/controllers/maintenances_controller.rb:
```diff
@@ -1,2 +1,69 @@
+  def destroy
+    @maintenance.destroy
+
+    redirect_to vehicle_maintenances_path(params[:vehicle_id])
```
Using the added before action to use the instance variable
```suggestion
    redirect_to vehicle_maintenances_path(@vehicle)
```

JPS09 reviewed (changes requested):
Great job on this CRUD, I will let you check out my suggestions which are mainly about using the instance variables instead of digging into the params :)

sbriclot committed:
Use vehicle/maintenance object instances when they exist

sbriclot commented:
You're right.
I removed the calls to methods **set_vehicle** and **set_maintenance** from everywhere, and I just restored them when they are mandatory (in render, not in redirect_to), and I forgot to replace params[] by the objects.
I fixed the issue in the methods **create** and **update**.
But I didn't update the method **destroy**, I don't need to select the object instance using the params[] if I can have the same result by using directly the params[] in the command redirect_to.

JPS09 reviewed (changes requested):
Great ! Thank you for these changes. I think that there is a few things that could improve what you did, please check them out below :)

sbriclot committed:
Also use objects instances in views

sbriclot commented:
My mistake, of course, the same changes should have been also made in the views.
Done now, thank you for the reviews.
I let you check it one more time.

JPS09 committed:
Update app/views/maintenance_plans/index.html.erb

JPS09 commented:
> 
> 
> My mistake, of course, the same changes should have been also made in the views.
> Done now, thank you for the reviews.
> I let you check it one more time.

Thank you for having the patience of making all these changes :) I committed a suggestion that was along the lines of what you already did so that you won't have to do it again ^^ It's ok for me, it can be merged now

sbriclot commented:
Thank you for the last commit, I merge the branch now.
